### PR TITLE
8212155: Race condition when posting dynamic_code_generated event leads to JVM crash

### PR DESCRIPTION
**Pull request by lmesnik**
Fixed a race condition between posting and enabling DynamicCodeGenerated event.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8212155](https://bugs.openjdk.java.net/browse/JDK-8212155): Race condition when posting dynamic_code_generated event leads to JVM crash

### Reviewers
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to c650f446379d5d9a24182a6f913b57d39c436dd3
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4331/head:pull/4331` \
`$ git checkout pull/4331`

Update a local copy of the PR: \
`$ git checkout pull/4331` \
`$ git pull https://git.openjdk.java.net/jdk pull/4331/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4331`

View PR using the GUI difftool: \
`$ git pr show -t 4331`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4331.diff">https://git.openjdk.java.net/jdk/pull/4331.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiExport.cpp [hunk 1]
```diff
@@ -2289,13 +2289,15 @@ void JvmtiExport::post_dynamic_code_generated_while_holding_locks(const char* na
   // register the stub with the current dynamic code event collector
   // Cannot take safepoint here so do not use state_for to get
   // jvmti thread state.
+  // The collector and/or state might be NULL if JvmtiDynamicCodeEventCollector has been initialized
+  // while JVMTI_EVENT_DYNAMIC_CODE_GENERATED was disabled
```
Comment (sspitsyn) [inline src/hotspot/share/prims/jvmtiExport.cpp:2293]: Could you, reballance this comment a little bit and add a dot at the end?

**Review by sspitsyn — APPROVED** [review]
Hi Leonid,
The fix looks good to me.
Thanks,
Serguei

**Review by sspitsyn — APPROVED** [review]

File: test/hotspot/jtreg/serviceability/jvmti/DynamicCodeGenerated/libDynamicCodeGenerated.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+jvmtiEnv* jvmti;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+JNIEXPORT
+void JNICALL Java_DynamicCodeGeneratedTest_changeEventNotificationMode(JNIEnv* jni, jclass cls) {
+  while(true) {
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/serviceability/jvmti/DynamicCodeGenerated/libDynamicCodeGenerated.cpp:35]: nit - please add space before '('

File: test/hotspot/jtreg/serviceability/jvmti/DynamicCodeGenerated/libDynamicCodeGenerated.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include <string.h>
+#include <jvmti.h>
+
+jvmtiEnv* jvmti;
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/serviceability/jvmti/DynamicCodeGenerated/libDynamicCodeGenerated.cpp:27]: Should this be 'static'?

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.

I just have minor comments.

**Comment by dcubed-ojdk** [comment]
Just curious: what's the execution times for the new test on the Mach5 platforms?

File: test/hotspot/jtreg/serviceability/jvmti/DynamicCodeGenerated/DynamicCodeGeneratedTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        });
+        t.setDaemon(true);
+        t.start();
+
+        for (int i = 0; i < 10_000; i++) {
+            new Thread(() -> {
+                String result = "string" + System.currentTimeMillis();
+                Reference.reachabilityFence(result);
+            }).start();
+        }
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/serviceability/jvmti/DynamicCodeGenerated/DynamicCodeGeneratedTest.java:55]: I just noticed no `join()` calls to clean up these threads.
Does this mean we'll have 10,000 thread objects waiting around
until the end of the program?

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.

File: test/hotspot/jtreg/serviceability/jvmti/DynamicCodeGenerated/DynamicCodeGeneratedTest.java (hunk 4 above)
Comment (lmesnik) [reply test/hotspot/jtreg/serviceability/jvmti/DynamicCodeGenerated/DynamicCodeGeneratedTest.java:55]: Yes, we don't care about thread completion. Just start new threads while the first ones are completed.  I  reduced the number of threads to 2000. It is still enough to reproduce the crash. However, 2,000 thread doesn't harm any system. I checked in Mach5.

**Review by lmesnik — COMMENTED** [review]
> Just curious: what's the execution times for the new test on the Mach5 platforms?

It takes ~3 seconds to complete the test in Mach5.

**Pull request closed without merging**